RandLAPACK/drivers/rl_revd2.hh

```diff
@@ -0,0 +1,240 @@
+        std::vector<T> R;
+        std::vector<T> S;
+
+        std::vector<T> A_cpy;
+};
+
+// -----------------------------------------------------------------------------
+template <typename T>
+int REVD2<T>::call(
+        int64_t m, // m = n
+        std::vector<T>& A,
+        int64_t& k,
+        std::vector<T>& V,
+        std::vector<T>& E
+){
+    T err = 0;
+    while(1) {
+        T* A_dat = A.data();
+        T* V_dat = util::upsize(m * k, V);
+        util::upsize(k, E);
+        T* Y_dat = util::upsize(m * k, this->Y);
+        T* Omega_dat = util::upsize(m * k, this->Omega);
+        T* R_dat = util::upsize(k * k, this->R);
+        T* S_dat = util::upsize(k * k, this->S);
+
+        // Construnct a sketching operator
+        // If CholeskyQR is used for stab/orth here, RF fails
+        if(this->RF_Obj.call(m, m, A, k, this->Omega)) {
+            printf("RF/RS FAILED\n");
+            return 2;
+        }
+
+        // Y = A * S
+        blas::gemm(Layout::ColMajor, Op::NoTrans, Op::NoTrans, m, k, m, 1.0, A_dat, m, Omega_dat, m, 0.0, Y_dat, m);
+
+        T nu = std::sqrt(m) * std::numeric_limits<double>::epsilon() * lapack::lange(lapack::Norm::Fro, m, k, Y_dat, m);
+
+        // We need Y = Y + vS
+        // We further need R = chol(S' Y)
+        // Solve this as R = chol(S' Y + vS'S)
+        // Compute vS'S - syrk only computes the lower triangular part. Need full.
+        blas::syrk(Layout::ColMajor, Uplo::Lower, Op::Trans, k, m, nu, Omega_dat, m, 0.0, R_dat, k);
+        // Fill the upper triangular part of S'S
+        for(int i = 1; i < k; ++i)
+            blas::copy(k - i, R_dat + i + ((i-1) * k), 1, R_dat + (i - 1) + (i * k), k);
+        // Compute S' Y + vS'S
+        blas::gemm(Layout::ColMajor, Op::Trans, Op::NoTrans, k, k, m, 1.0, Omega_dat, m, Y_dat, m, 1.0, R_dat, k);
+
+        // Compute R = chol(S' Y + vS'S)
+        // Looks like if POTRF gets passed a non-triangular matrix, it will also output a non-triangular one
+        if(lapack::potrf(Uplo::Upper, k, R_dat, k)) {
+            printf("CHOLESKY FACTORIZATION FAILED\n");
+            return 1;
+        }
+        RandLAPACK::util::get_U(k, k, R);
+
+        // B = Y(R')^-1 - need to transpose R
+        blas::trsm(Layout::ColMajor, Side::Right, Uplo::Upper, Op::NoTrans, Diag::NonUnit, m, k, 1.0, R_dat, k, Y_dat, m);
+
+        //[V, S, ~] = SVD(B)
+        // Although we don't need the right singular vectors, we need to give space for those.
+        // Use R as a buffer for that.
+        lapack::gesdd(Job::SomeVec, m, k, Y_dat, m, S_dat, V_dat, m, R_dat, k);
+
+        // E = diag(S^2)
+        T buf;
+        int64_t r = 0;
+        int i;
+        for(i = 0; i < k; ++i) {
+            buf = std::pow(S[i], 2);
+            E[i] = buf;
+            // r = number of entries in E that are greater than v
+            if(buf > nu)
+                ++r;
+        }
+
+        // Undo regularlization
+        for(i = 0; i < r; ++i)
+            E[i] -=nu;
```

Comment:
This might make some entries of ``E`` negative. Make sure that the code which follows accounts for that possibility.

Reply:
Fixed.